components/veml7700/veml7700.c

```diff
@@ -0,0 +1,237 @@
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_write(dev, &command_code, 1, &data, 2));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+static uint32_t resolution(veml7700_config_t *config)
+{
+    CHECK_ARG(config);
+
+    uint32_t resolution = VEML7700_RESOLUTION_800MS_IT_GAIN_2;
+    switch (config->gain)
+    {
+        case VEML7700_GAIN_1:
+            resolution = resolution * 2;
+            break;
+        case VEML7700_GAIN_DIV_4:
+            resolution = resolution * 8;
+            break;
+        case VEML7700_GAIN_DIV_8:
+            resolution = resolution * 16;
+            break;
+    }
+    switch (config->integration_time)
+    {
+        case VEML7700_INTEGRATION_TIME_400MS:
+            resolution = resolution * 2;
+            break;
+        case VEML7700_INTEGRATION_TIME_200MS:
+            resolution = resolution * 4;
+            break;
+        case VEML7700_INTEGRATION_TIME_100MS:
+            resolution = resolution * 8;
+            break;
+        case VEML7700_INTEGRATION_TIME_50MS:
+            resolution = resolution * 16;
+            break;
+        case VEML7700_INTEGRATION_TIME_25MS:
+            resolution = resolution * 32;
+            break;
+    }
+    return resolution;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+esp_err_t veml7700_init_desc(i2c_dev_t *dev, i2c_port_t port, gpio_num_t sda_gpio, gpio_num_t scl_gpio)
+{
+    CHECK_ARG(dev);
+
+    dev->port = port;
+    dev->addr = VEML7700_I2C_ADDR;
+    dev->cfg.sda_io_num = sda_gpio;
+    dev->cfg.scl_io_num = scl_gpio;
+#if HELPER_TARGET_IS_ESP32
+    dev->cfg.master.clk_speed = I2C_FREQ_HZ;
+#endif
+
+    return i2c_dev_create_mutex(dev);
+}
+
+esp_err_t veml7700_free_desc(i2c_dev_t *dev)
+{
+    CHECK_ARG(dev);
+
+    return i2c_dev_delete_mutex(dev);
+}
+
+esp_err_t veml7700_probe(i2c_dev_t *dev)
+{
+    CHECK_ARG(dev);
+
+    /* use write request since read request causes a timeout;
+     * just doing a read is not intended to use by the chip,
+     * it is waiting for a command code
+     */
+    return i2c_dev_probe(dev, I2C_DEV_WRITE);
+}
+
+esp_err_t veml7700_config(i2c_dev_t *dev, veml7700_config_t *config)
+{
+    CHECK_ARG(dev);
+    CHECK_ARG(config);
+
+    uint16_t config_data = 0;
+    config_data |= config->gain << VEML7700_GAIN_SHIFT;
+    config_data |= config->integration_time << VEML7700_INTEGRATION_TIME_SHIFT;
+    config_data |= config->persistence_protect << VEML7700_PERSISTENCE_PROTECTION_SHIFT;
+    config_data |= config->interrupt_enable << VEML7700_INTERRUPT_ENABLE_SHIFT;
+    config_data |= config->shutdown << VEML7700_SHUTDOWN_SHIFT;
+
+    uint16_t power_saving_data = 0;
+    power_saving_data |= config->power_saving_mode << VEML7700_POWER_SAVING_MODE_SHIFT;
+    power_saving_data |= config->power_saving_enable << VEML7700_POWER_SAVING_MODE_ENABLE_SHIFT;
+
+    CHECK(write_port(dev, VEML7700_COMMAND_CODE_ALS_CONF_0, config_data));
```

Comment:
Not quite correct work with the device mutex.

As I understand it, there is one atomic operation from the point of view of the programmer: setting the device settings. Accordingly, instead of taking / releasing the device mutex for each individual I2C transaction, you need to take it once before all writes are performed and release it once at the end (or in case of an error).
```C
I2C_DEV_TAKE_MUTEX(dev);
write_port(...);
write_port(...);
write_port(...);
I2C_DEV_GIVE_MUTEX(dev);
```